pkg/apis/experimental/types.go

```diff
@@ -508,35 +509,56 @@ type IngressStatus struct {
 	LoadBalancer api.LoadBalancerStatus `json:"loadBalancer,omitempty"`
 }
 
-// IngressRule represents the rules mapping the paths under a specified host to the related backend services.
-type IngressRule struct {
+// IngressHost represents the rules mapping the paths under a specified host to
+// the related backend services.
+type IngressHost struct {
 	// Host is the fully qualified domain name of a network host, or its IP
 	// address as a set of four decimal digit groups separated by ".".
-	// Conforms to RFC 1738.
+	// Conforms to RFC 1738. Incoming requests are matched against the Host
+	// before the IngressRule.
 	Host string `json:"host,omitempty"`
+	// IngressRule represents a rule to route requests for this IngressHost.
+	IngressRule `json:",inline"`
+}
+
+// IngressRule represents a rule to apply against incoming requests. If the
+// rule is satisfied, the request is routed to the specified backend.
+type IngressRule struct {
+	// Currently mixing different types of rules in a single Ingress is
+	// disallowed, so exactly one of the following must be set.
+	HTTP *HTTPIngressRule `json:"http"`
+}
 
-	// Paths describe a list of load-balancer rules under the specified host.
-	Paths []IngressPath `json:"paths"`
+// HTTPIngressRule is a list of http selectors pointing to IngressBackends.
+// In the example: http://<host>:<port>/<path>?<searchpart> -> IngressBackend
+// where parts of the url correspond to RFC 1738, this resource will be used
+// to match against everything after the last '/'.
+type HTTPIngressRule struct {
+	// A collection of paths that map requests to IngressBackends.
+	Paths []HTTPIngressPath `json:"paths"`
 }
 
 // IngressPath associates a path regex with an IngressBackend.
 // Incoming urls matching the Path are forwarded to the Backend.
-type IngressPath struct {
-	// Path is a regex matched against the url of an incoming request.
+type HTTPIngressPath struct {
+	// Path is a extended POSIX regex as defined by IEEE Std 1003.1,
+	// (i.e this follows the egrep/unix syntax, not the perl syntax)
+	// matched against the path of an incoming request. Currently it can
+	// contain characters disallowed from the conventional "path"
+	// part of a URL as defined by RFC 3986. Paths must begin with
+	// a '/'.
 	Path string `json:"path,omitempty"`
 
-	// Define the referenced service endpoint which the traffic will be forwarded to.
+	// Define the referenced service endpoint which the traffic will be
+	// forwarded to.
 	Backend IngressBackend `json:"backend"`
 }
 
-// IngressBackend describes all endpoints for a given Service, port and protocol.
+// IngressBackend describes all endpoints for a given Service and port.
 type IngressBackend struct {
 	// Specifies the referenced service.
 	ServiceRef api.LocalObjectReference `json:"serviceRef"`
```

Comment:
Ref = another object in api (regardless of typeof(ref)), that seems clear to me from a usability perspective. I initially had it as `ServiceName string`, but in theory we could grow a local object ref, or create another object ref to contain more information (like object meta?) and that would be useful here. 

Don't feel strongly eitherway, so defer back to @thockin for final call.